src/router_core/connections.c

```diff
@@ -273,8 +273,14 @@ void qdr_record_link_credit(qdr_core_t *core, qdr_link_t *link)
 void qdr_close_connection_CT(qdr_core_t *core, qdr_connection_t  *conn)
 {
     sys_mutex_lock(&conn->work_lock);
-    conn->closed = true;
-    conn->error  = qdr_error(QD_AMQP_COND_CONNECTION_FORCED, "Connection forced-closed by management request");
+    //
+    // In some cases, this function can be called on an already closed connection in which case
+    // we don't want to leak a qdr_error_t object
+    //
+    if (!conn->closed) {
+        conn->closed = true;
```

Comment:
In here I'd try setting a local flag to call the activate once the lock is dropped.  That would ensure that the close handler is not invoked twice.

Try that.  Does it break anything?

Reply:
I used a local flag to call activate outside the lock and it does the job, does not break anything. I am running the http2 and http2_tls tests in a loop with -j4 and will run it for some more time to see if it catches anything